lib/UBOTypeTransformPass.cpp

```diff
@@ -0,0 +1,520 @@
+    if (auto old_init = GV->getInitializer()) {
+      initializer = RebuildConstant(old_init, replacement_type, M);
+    }
+    // Recreate the global variable.
+    GlobalVariable *replacement = new GlobalVariable(
+        M, replacement_type, GV->isConstant(), GV->getLinkage(), initializer,
+        GV->getName(), /*InsertBefore=*/nullptr, GV->getThreadLocalMode(),
+        GV->getType()->getPointerAddressSpace());
+    remapped_globals_[GV] = replacement;
+    replacement->copyMetadata(GV, 0);
+  }
+
+  return changed;
+}
+
+bool UBOTypeTransformPass::RemapUser(User *user, Module &M) {
+  if (isa<ConstantData>(user) || isa<ConstantAggregate>(user)) {
+    return RemapConstant(cast<Constant>(user), M);
+  }
+
+  bool changed = RemapValue(user, M);
+
+  for (Use &use : user->operands()) {
+    User *operand_user = use.getUser();
+    if (!operand_user) {
+      changed |= RemapValue(use.get(), M);
+    } else if (!isa<Instruction>(operand_user) &&
+               !isa<GlobalValue>(operand_user) &&
+               !isa<Argument>(operand_user)) {
+      // Keep mutating to handle constant expressions.
+      changed |= RemapUser(operand_user, M);
+      // If this was a constant that got rebuilt, update the operand.
+      if (auto *constant = dyn_cast<Constant>(operand_user)) {
+        auto iter = remapped_globals_.find(constant);
+        if (iter != remapped_globals_.end()) {
+          use.set(iter->second);
+          changed = true;
+        }
+      }
+    }
+  }
+
+  return changed;
+}
+
+bool UBOTypeTransformPass::RemapValue(Value *value, Module &M) {
+  Type *remapped = MapType(value->getType(), M);
+  if (remapped == value->getType())
+    return false;
+
+  value->mutateType(remapped);
+  return true;
+}
+
+bool UBOTypeTransformPass::RemapConstant(Constant *constant, Module &M) {
+  // Rebuild the constant.
+  Type *remapped = MapType(constant->getType(), M);
+  if (remapped == constant->getType())
+    return false;
+
+  RebuildConstant(constant, remapped, M);
+  return true;
+}
+
+Constant *UBOTypeTransformPass::RebuildConstant(Constant *constant,
+                                                Type *remapped_ty, Module &M) {
+  if (constant->getType() == remapped_ty)
+    return constant;
+
+  // Check whether this constant has been rebuilt already.
+  auto iter = remapped_globals_.find(constant);
+  if (iter != remapped_globals_.end())
+    return iter->second;
+
+  if (constant->isZeroValue()) {
+    Constant *null_constant = Constant::getNullValue(remapped_ty);
+    remapped_globals_[constant] = null_constant;
+    return null_constant;
+  } else if (isa<UndefValue>(constant)) {
+    // This case should catch the padding transformations since the padding
+    // can't be initialized.
+    Constant *undef_constant = UndefValue::get(remapped_ty);
+    remapped_globals_[constant] = undef_constant;
+    return undef_constant;
+  } else if (auto *agg_constant = dyn_cast<ConstantAggregate>(constant)) {
```

Comment:
I wonder if you'll also need a ConstantDataSequential case. That only inherits from ConstantData

I guess we'll find out if you hit the llvm_unreachable

Reply:
ok for now since this is only turned on under flag.